tests/e2e/unified_test.go

```diff
@@ -603,6 +603,73 @@ func runScenario(ctx context.Context, t *testing.T, testPause bool, fixture reso
 					addReplacement("primary.generateTime", "2024-04-01T12:34:56.123456Z")
+					jsonMutators = append(jsonMutators, func(obj map[string]any) {
+						if val, found, _ := unstructured.NestedString(obj, "name"); found {
+							if strings.Contains(val, "clusters/alloydb") ||
+								strings.Contains(val, "instances/alloydb") ||
+								strings.Contains(val, "backups/alloydb") {
+
+								// Explicitly set `reconciling` to `false`.
+								if _, found, _ := unstructured.NestedBool(obj, "reconciling"); !found {
+									if err := unstructured.SetNestedField(obj, false, "reconciling"); err != nil {
+										t.Fatal(err)
+									}
+								}
+
+								// Explicitly set `automatedBackupPolicy.enabled` in response to `false`.
+								if _, found, _ := unstructured.NestedMap(obj, "automatedBackupPolicy"); found {
+									if _, found, _ := unstructured.NestedBool(obj, "automatedBackupPolicy", "enabled"); !found {
+										if err := unstructured.SetNestedField(obj, false, "automatedBackupPolicy", "enabled"); err != nil {
+											t.Fatal(err)
+										}
+									}
+								}
+							}
+						}
+					})
+					// Boolean fields in LRO are omitted when false so we need
+					// to add them back.
+					jsonMutators = append(jsonMutators, func(obj map[string]any) {
+						if _, found, _ := unstructured.NestedMap(obj, "metadata"); found {
+							if val, found, err := unstructured.NestedString(obj, "metadata", "@type"); err == nil && found && val == "type.googleapis.com/google.cloud.alloydb.v1beta.OperationMetadata" {
+								if _, found, err := unstructured.NestedString(obj, "done"); err == nil && !found {
+									// Explicitly set `done` to `false`.
+									if err := unstructured.SetNestedField(obj, false, "done"); err != nil {
+										t.Fatal(err)
+									}
+								}
+
+								if _, found, err := unstructured.NestedString(obj, "metadata", "requestedCancellation"); err == nil && !found {
+									// Explicitly set `metadata.requestedCancellation` to `false`.
+									if err := unstructured.SetNestedField(obj, false, "metadata", "requestedCancellation"); err != nil {
+										t.Fatal(err)
+									}
+								}
+
+								if _, found, _ := unstructured.NestedMap(obj, "response"); found {
+									if val, found, _ := unstructured.NestedString(obj, "response", "@type"); found &&
+										val == "type.googleapis.com/google.cloud.alloydb.v1beta.Cluster" ||
+										val == "type.googleapis.com/google.cloud.alloydb.v1beta.Instance" ||
+										val == "type.googleapis.com/google.cloud.alloydb.v1beta.Backup" {
+										// Explicitly set `reconciling` in response to `false`.
+										if _, found, _ := unstructured.NestedBool(obj, "response", "reconciling"); !found {
+											if err := unstructured.SetNestedField(obj, false, "response", "reconciling"); err != nil {
+												t.Fatal(err)
+											}
+										}
+										// Explicitly set `automatedBackupPolicy.enabled` in response to `false`.
```

Comment:
Can we do this in the Mock service level? My concern is that this could override other resources and make the triaging very hard.

Reply:
I tried and it didn't work. AFAIK, the empty values of the primitive typed fields are somehow dropped in the http log recording for mock. Possibly a gap in the dependent lib? We should look into it when it becomes more problematic. I think other folks ran into the same issue as well.

Regarding the concern about unexpected override, the code I added checks the returned resource name to ensure they are responses for AlloyDB resources. So it's unlikely to override non-AlloyDB resources.

Reply:
I think it is a proto3 explicit optional, so that's relatively unusual.

If we set it explicitly to false (`.Enabled = PtrTo(false)`), I would hope the protojson marshal code would understand but I guess it doesn't....

Reply:
I commented out both blocks that  `// Explicitly set `automatedBackupPolicy.enabled` in response to `false`.` (this one and the LRO one).  I then ran `hack/compare-mock fixtures/basicalloydbinstance` and it passed.  I also tried `hack/compare-mock fixtures/basicalloydbcluster` and was able to recreate the issue.  The thing is that the diff happens way down the log, and it actually happens after a patch to automatedBackupPolicy.

So I actually think that the problem is that we should default automatedBackupPolicy.enabled to PtrTo(false), and we aren't doing that after a patch.

Something like this:

```
~/kcc/k8s-config-connector  (202411-alloydb-mockgcp) > git diff
diff --git a/mockgcp/mockalloydb/cluster.go b/mockgcp/mockalloydb/cluster.go
index 362aa7847..fb230305a 100644
--- a/mockgcp/mockalloydb/cluster.go
+++ b/mockgcp/mockalloydb/cluster.go
@@ -65,11 +65,11 @@ func setClusterFields(name *clusterName, obj *pb.Cluster) {
                        BackupWindow: direct.PtrTo(duration.Duration{Seconds: 3600}),
                        // Defaults to true when unset, which is different from the value in
                        // the default policy.
-                       Enabled:  direct.PtrTo(false),
+                       Enabled:  PtrTo(false),
                        Location: name.Location,
                        Retention: &pb.AutomatedBackupPolicy_TimeBasedRetention_{
                                TimeBasedRetention: &pb.AutomatedBackupPolicy_TimeBasedRetention{
-                                       RetentionPeriod: direct.PtrTo(duration.Duration{Seconds: 1209600}),
+                                       RetentionPeriod: PtrTo(duration.Duration{Seconds: 1209600}),
                                },
                        },
                        Schedule: &pb.AutomatedBackupPolicy_WeeklySchedule_{
@@ -240,6 +240,10 @@ func (s *AlloyDBAdminV1) UpdateCluster(ctx context.Context, req *pb.UpdateCluste
                obj.ContinuousBackupInfo.EnabledTime = nil
        }
 
+       if obj.AutomatedBackupPolicy != nil && obj.AutomatedBackupPolicy.Enabled == nil {
+               obj.AutomatedBackupPolicy.Enabled = PtrTo(false)
+       }
+
        if err := s.storage.Update(ctx, fqn, obj); err != nil {
                return nil, err
        }

```


(We also should not be using direct.PtrTo from mockgcp, because then we introduce a circular dependency into the modules which makes updating them tricky.  In other services we've created a utils.go in each service.  We could always have a similar set of utils for mockgcp - we do have common so maybe that's a natural place)

Reply:
Thank you for all the details, @justinsb ! Yes, the optional fields (e.g. `automatedBackupPolicy.enabled`) should have their values recorded in the http log as is, and the missing part here is to set the value of `automatedBackupPolicy.enabled` explicitly again in the mock Update.

@yuwenma Didn't realize you are calling out this specific field. Should be updated now. My original comment is about other fields that need to have `false` explicitly set in the normalizer.